text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,180 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Specification
+
+Stack image creators may add the following executables to their run and/or build images:
+
+```
+/cnb/image/build/extend (pkg-cache-dir) (build-metadata-toml-file) | cwd: /
+/cnb/image/run/extend (pkg-cache-dir) (run-metadata-toml-file) | cwd: /
+```
+* executes on base run or build images
+* writes metadata to `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml`
+* on subsequent builds, lookup latest upstream base image digest:
+  * if the digest is different than extended image's base, the new upstream digest is extended
+  * if the digest is the same, proceed to status
+
+```
+/cnb/image/build/status (pkg-cache-dir) (build-metadata-toml-file) | cwd: /
+/cnb/image/run/status (pkg-cache-dir) (run-metadata-toml-file) | cwd: /
+```
+* executes on run or build base images
+* metadata arg is `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml`
+* exit status 100 = existing base image digest is extended 
+* exit status 0 = does not need extension
+* exit status other = unknown
+
+The status executable may be invoked to determine whether updates are necessary.
+
+## UX
+
+### Creating an Extended Builder
+
+`pack create-builder sclevine/builder -b builder.toml`
+
+builder.toml:
+```toml
+[stack]
+id = "io.buildpacks.stacks.bionic"
+build-image = "example.com/build"
+run-image = "example.com/run"
+run-image-mirrors = ["example.org/run"]
+
+[extend]
+run-image = "sclevine/run"
+run-image-mirrors = ["scl.sh/run"]
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[extend.build] # build-metadata-toml-file (stored on builder image metadata)
+packages = ["git"]
+```
+
+Behavior: creates a new builder with additional packages as well as a new run image (`sclevine/run`) with additional packages.
+
+If `extend.run-image` is not specified, `extend.run` is stored on the builder metadata and used to dynamically extend the run image on `pack build`.
+
+### Extended an Existing Builder
+
+`pack extend-builder sclevine/builder -e extend.toml`
+
+extend.toml:
+```toml
+[builder]
+image = "example.com/builder"
+
+[extend]
+run-image = "sclevine/run"
+run-image-mirrors = ["scl.sh/run"]
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[extend.build] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+```
+
+Behavior: creates a new version of an existing builder with additional packages as well as a new run image (`sclevine/run`) with additional packages.
+
+If `extend.run-image` is not specified, `extend.run` is stored on the new builder metadata and used to dynamically extend the run image on `pack build`.
+
+### Building an App with Additional Packages
+
+`pack build sclevine/myapp`
+
+project.toml:
+```toml
+...
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
```

Comment:
`[extend]` is a little awkward as a top level key if we adopt an `[extensions]` key 😬

Reply:
I don't feel strongly about the naming, and placing the section under each image seems reasonable.